device/send.go

```diff
@@ -305,26 +350,56 @@ top:
 	}
 
 	for {
+		var elemsOOO *[]*QueueOutboundElement
 		select {
-		case elem := <-peer.queue.staged:
-			elem.peer = peer
-			elem.nonce = keypair.sendNonce.Add(1) - 1
-			if elem.nonce >= RejectAfterMessages {
-				keypair.sendNonce.Store(RejectAfterMessages)
-				peer.StagePacket(elem) // XXX: Out of order, but we can't front-load go chans
-				goto top
+		case elems := <-peer.queue.staged:
+			i := 0
+			for _, elem := range *elems {
+				elem.peer = peer
+				elem.nonce = keypair.sendNonce.Add(1) - 1
+				if elem.nonce >= RejectAfterMessages {
+					peer.device.log.Errorf("reordering message due to invalid nonce for peer %s", peer)
```

Comment:
Do we really need to error message over this? This is the same gotcha that was there before, with nothing changed about it right? So if you want to start warning about it, that probably needs independent justification and hence a separate commit.

Reply:
Ah, we were studying how often this comes up in practice, I think this leaked through. Removed.

Reply:
More broadly, this is also an area we're eager to discuss sometime - other possible arrangements for the IO path and channels, but as you note elsewhere, there are some details of the IO path semantics that are relatively closely tied to the protocol - so it's definitely more of a discussion first than any kind of ready or concrete proposal.

Reply:
Sure, happy to talk about this whenever. Maybe we should have a phone call sometime soon so we can more casually toss ideas around about that.